src/audio/base_fw.c

```diff
@@ -228,12 +234,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
 	return 0;
 }
 
+static uint32_t basefw_set_system_time(uint32_t param_id,
+				       bool first_block,
+				       bool last_block,
+				       uint32_t data_offset,
+				       char *data)
+{
+	if (!(first_block && last_block))
+		return IPC4_INVALID_REQUEST;
+
+	global_system_time_info.host_time.val_l = (*(struct ipc4_system_time *)(data)).val_l;
+	global_system_time_info.host_time.val_u = (*(struct ipc4_system_time *)(data)).val_u;
+	uint64_t current_dsp_time = sof_cycle_get_64();
+
+	global_system_time_info.dsp_time.val_l = (uint32_t)(current_dsp_time);
+	global_system_time_info.dsp_time.val_u = (uint32_t)(current_dsp_time >> 32);
+	return IPC4_SUCCESS;
+}
+
+static uint32_t basefw_get_system_time(uint32_t *data_offset, char *data)
+{
+	struct ipc4_system_time *system_time = (struct ipc4_system_time *)data;
+
+	system_time->val_l = global_system_time_info.host_time.val_l;
+	system_time->val_u = global_system_time_info.host_time.val_u;
+	*data_offset = sizeof(struct ipc4_system_time);
+	return IPC4_SUCCESS;
+}
+
+static uint32_t basefw_get_ext_system_time(uint32_t *data_offset, char *data)
+{
+	uint64_t host_time = ((uint64_t)global_system_time_info.host_time.val_u << 32)
+				| (uint64_t)global_system_time_info.host_time.val_l;
+	uint64_t dsp_time = ((uint64_t)global_system_time_info.dsp_time.val_u << 32)
+				| (uint64_t)global_system_time_info.dsp_time.val_l;
+	(void)host_time;
+	(void)dsp_time;
+
+#if CONFIG_ACE_V1X_ART_COUNTER && CONFIG_ACE_V1X_RTC_COUNTER
+	struct ipc4_ext_system_time *ext_system_time = (struct ipc4_ext_system_time *)(data);
+	struct ipc4_ext_system_time ext_system_time_data = {0};
+
+	uint64_t hh = 0;
+	uint64_t wallclk = 0;
+	uint64_t rtc = 0;
+
+	const struct device *dev;
+	const struct counter_driver_api *api;
+
+	dev = device_get_binding(DT_LABEL(DT_NODELABEL(ace_art_counter)));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	api = (const struct counter_driver_api *)dev->api;
+	api->get_value_64(dev, &hh);
+	wallclk = sof_cycle_get_64();
+	ext_system_time_data.hh_l = (uint32_t)hh;
+	ext_system_time_data.hh_u = (uint32_t)(hh >> 32);
+	uint64_t delta = (wallclk - dsp_time) / (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / 1000000);
+
+	uint64_t new_host_time = (host_time + delta);
+
+	ext_system_time_data.utc_l = (uint32_t)new_host_time;
+	ext_system_time_data.utc_u = (uint32_t)(new_host_time >> 32);
+
+	dev = device_get_binding(DT_LABEL(DT_NODELABEL(ace_rtc_counter)));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	api = (const struct counter_driver_api *)dev->api;
+	api->get_value_64(dev, &rtc);
+	ext_system_time_data.rtc_l = (uint32_t)rtc;
+	ext_system_time_data.rtc_u = (uint32_t)(rtc >> 32);
+
+	memcpy_s(ext_system_time, sizeof(ext_system_time), &ext_system_time_data,
+		 sizeof(ext_system_time));
+	*data_offset = sizeof(struct ipc4_ext_system_time);
+	return IPC4_SUCCESS;
+#endif
+	return IPC4_FAILURE;
```

Comment:
wrong return, this is general (unknown) failure

Reply:
And breaks integration with kernel. If we report an error, the kernel will refuse to work with the FW. So we need to silently ignore this if unimplemented.